RFS/src/main/java/com/rfs/common/DocumentReindexer.java

```diff
@@ -1,32 +1,124 @@
 package com.rfs.common;
 
+import java.time.Duration;
+import java.util.Base64;
+import java.util.List;
+
+import io.netty.buffer.Unpooled;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.lucene.document.Document;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+import reactor.util.retry.Retry;
 
 
 public class DocumentReindexer {
     private static final Logger logger = LogManager.getLogger(DocumentReindexer.class);
+    private static final int MAX_BATCH_SIZE = 1000; // Arbitrarily chosen
+
+    public static void reindex(String indexName, Flux<Document> documentStream, ConnectionDetails targetConnection) throws Exception {
+        String targetUrl = "/" + indexName + "/_bulk";
+        HttpClient client = HttpClient.create()
+            .host(targetConnection.hostName)
+            .port(targetConnection.port)
+            .headers(h -> {
+                h.set("Content-Type", "application/json");
+                if (targetConnection.authType == ConnectionDetails.AuthType.BASIC) {
+                    String credentials = targetConnection.username + ":" + targetConnection.password;
+                    String encodedCredentials = Base64.getEncoder().encodeToString(credentials.getBytes());
+                    h.set("Authorization", "Basic " + encodedCredentials);
+                }
+            });
+
+        documentStream
+            .map(DocumentReindexer::convertDocumentToBulkSection)  // Convert each Document to part of a bulk operation
+            .buffer(MAX_BATCH_SIZE) // Collect until you hit the batch size
+            .map(DocumentReindexer::convertToBulkJson)  // Assemble the bulk request body from the parts
+            .flatMap(bulkJson -> sendBulkRequest(client, targetUrl, bulkJson)) // Send the request
+            .retryWhen(Retry.backoff(3, Duration.ofSeconds(1)).maxBackoff(Duration.ofSeconds(5)))
+            .subscribe(
+                response -> logger.info("Batch uploaded successfully"),
+                error -> logger.error("Failed to upload batch", error)
+            );
+    }
 
-    public static void reindex(String indexName, Document document, ConnectionDetails targetConnection) throws Exception {        
-        // Get the document details
+    private static String convertDocumentToBulkSection(Document document) {
         String id = Uid.decodeId(document.getBinaryValue("_id").bytes);
         String source = document.getBinaryValue("_source").utf8ToString();
+        String action = "{\"index\": {\"_id\": \"" + id + "\"}}";
 
-        logger.info("Reindexing document - Index: " + indexName + ", Document ID: " + id);
+        return action + "\n" + source;
+    }
 
-        // Assemble the request details
-        String path = indexName + "/_doc/" + id;
-        String body = source;
+    private static String convertToBulkJson(List<String> bulkSections) {
```

Comment:
nit: the function name`...BulkJson` was a bit confusing since it's just delimited jsons, maybe `convertToBulkBody` or convertToDelimitedJsons`

Reply:
Sure, will do.